File: go/ai-service/internal/llm/factory.go
package llm

import (
	"fmt"

	gobreaker "github.com/sony/gobreaker/v2"
)

// NewClient creates an LLM Client based on the provider name.
// Supported providers: "ollama", "openai", "anthropic".
// The breaker is applied to ollama; other providers ignore it for now.
func NewClient(provider, baseURL, model, apiKey string, breaker *gobreaker.CircuitBreaker[any]) (Client, error) {
	switch provider {
	case "ollama":
		return NewOllamaClient(baseURL, model, breaker), nil
	case "openai":
		return NewOpenAIClient(baseURL, model, apiKey), nil
	case "anthropic":
		return NewAnthropicClient(model, apiKey), nil
	default:
		return nil, fmt.Errorf("unknown LLM provider: %q (use ollama, openai, or anthropic)", provider)
	}
}

File: services/shared/llm/factory.py
"""Factory functions to create LLM and embedding providers from env vars."""

from __future__ import annotations

from llm.anthropic_provider import AnthropicLLMProvider
from llm.ollama import OllamaEmbeddingProvider, OllamaLLMProvider
from llm.openai_compat import OpenAIEmbeddingProvider, OpenAILLMProvider


def get_llm_provider(
    provider: str,
    base_url: str,
    api_key: str,
    model: str,
) -> OllamaLLMProvider | OpenAILLMProvider | AnthropicLLMProvider:
    """Create an LLM provider based on the provider name."""
    if provider == "ollama":
        return OllamaLLMProvider(base_url=base_url, model=model)
    if provider == "openai":
        return OpenAILLMProvider(base_url=base_url, api_key=api_key, model=model)
    if provider == "anthropic":
        return AnthropicLLMProvider(api_key=api_key, model=model)
    raise ValueError(
        f"Unknown LLM provider: {provider!r}. Use 'ollama', 'openai', or 'anthropic'."
    )


def get_embedding_provider(
    provider: str,
    base_url: str,
    api_key: str,
    model: str,
) -> OllamaEmbeddingProvider | OpenAIEmbeddingProvider:
    """Create an embedding provider based on the provider name."""
    if provider == "ollama":
        return OllamaEmbeddingProvider(base_url=base_url, model=model)
    if provider == "openai":
        return OpenAIEmbeddingProvider(base_url=base_url, api_key=api_key, model=model)
    raise ValueError(
        f"Unknown embedding provider: {provider!r}. Use 'ollama' or 'openai'."
    )
